daceml/torch/module.py

```diff
@@ -396,21 +397,74 @@ def __sdfg__(self, *args):
             Using a PyTorch model in a DaceProgram requires that the model is initialized first.
             Either call this model using some inputs, or pass 'dummy_inputs' to the constructor.
             """)
+
+        for name, param in self._exported_parameters.items():
+            onnx_name = clean_onnx_name(name)
+            if param.requires_grad:
+                autodiff_library.Array.make_parameter(self.sdfg, onnx_name)
         return self.sdfg
 
+    def _add_gradient_buffers(self) -> List[str]:
+        """
+        Allocate gradient buffers for all parameters, and add their descriptors to the SDFG.
+
+        :return: a list of the sdfg array names of the gradient buffers
+        """
+
+        assert self.sdfg is not None
+        if hasattr(self, '_gradient_buffers'):
+            return self._gradient_buffers
+
+        buffers = []
+        for name, param in self._exported_parameters.items():
+            onnx_name = clean_onnx_name(name)
+            desc = self.sdfg.arrays[onnx_name]
+
+            if param.requires_grad:
+                # allocate gradient buffer
+                param.grad = torch.empty_like(param.data)
+
+                # add gradient buffer descriptor to sdfg
+                autodiff_library.Array.make_parameter(self.sdfg, onnx_name)
+                desc: autodiff_library.Array = self.sdfg.arrays[onnx_name]
+                grad_name = desc.add_gradient_buffer(self.sdfg, onnx_name)
+                grad_desc = self.sdfg.arrays[grad_name]
+                grad_desc.transient = False
+                buffers.append(grad_name)
+        self._gradient_buffers = buffers
+        return buffers
+
     def __sdfg_signature__(self):
         if self.dace_model is None:
             raise ValueError(
                 "Can't determine signature before SDFG is generated.")
         inputs = [clean_onnx_name(name) for name in self.dace_model.inputs]
+        grad_buffers = self._add_gradient_buffers()
+        inputs.extend(grad_buffers)
+
         return inputs, []
 
+    @staticmethod
+    def _tensor_from_param(param):
+        t = param.data
+        # for some reason doing .data on a Parameter kills the requires_grad flag
```

Comment:
it's because it is normally used in a no_grad environment to modify the data buffer itself.

Reply:
Will update this comment